samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java

```diff
@@ -181,12 +181,26 @@ private SystemStreamMetadata constructSystemStreamMetadata(
             .stream()
             .collect(Collectors.toMap(entry -> entry.getKey().getPartition(), entry -> {
                 List<IncomingMessageEnvelope> messages = entry.getValue();
-                String oldestOffset = messages.isEmpty() ? null : "0";
-                String newestOffset = messages.isEmpty() ? null : String.valueOf(messages.size() - 1);
-                String upcomingOffset = String.valueOf(messages.size());
-
-                return new SystemStreamMetadata.SystemStreamPartitionMetadata(oldestOffset, newestOffset, upcomingOffset);
-
+                Integer oldestOffset = null;
+                Integer newestOffset = null;
+                int upcomingOffset = 0;
```

Comment:
maybe simplify this to have it default to original values `0, messages.size() - 1, messages.size` and only modify in case of the last message being end of stream?

Reply:
IMO the definition of oldest and newest offset is such that they should be valid pointers to data. In the case where `messages` is empty or has only a single end of stream message, I think those values would be a misrepresentation.

`upcomingOffset` I am less sure of semantically but looking back again at the documentation In SystemStreamPartitionMetadata suggests it should be null as well in the case of an empty stream. It also doesn't really make sense in the context of bounded streams either (in HdfsSystemConsumer, it is returned as null).

I suggest we init upcomingOffset to null, remove it from being set where there is an end of stream message, and leave the other two as-is.

Thoughts?

Reply:
@mynameborat is this resolved after the latest diff?